src/rosetta/rosetta-ft-metadata-client.ts

```diff
@@ -0,0 +1,187 @@
+import {
+  ClarityType,
+  ClarityValue,
+  getAddressFromPrivateKey,
+  hexToCV,
+  makeRandomPrivKey,
+  TransactionVersion,
+  UIntCV,
+} from '@stacks/transactions';
+import { ChainID, getChainIDNetwork } from '../helpers';
+import { ReadOnlyContractCallResponse, StacksCoreRpcClient } from '../core-rpc/client';
+import { logger } from '../logger';
+import * as LRUCache from 'lru-cache';
+
+/** Fungible token metadata for Rosetta operations display */
+export interface RosettaFtMetadata {
+  symbol: string;
+  decimals: number;
+}
+
+interface RosettaFtContractCallParams {
+  address: string;
+  contractAddress: string;
+  contractName: string;
+  functionName: string;
+}
+
+enum RosettaTokenMetadataErrorMode {
+  /** Default mode. If a required token metadata is not found when it is needed for a response, the
+   * API will issue a warning. */
+  warning,
+  /** If a required token metadata is not found, the API will throw an error. */
+  error,
+}
+
+/**
+ * Determines the token metadata error handling mode based on .env values.
+ * @returns TokenMetadataMode
+ */
+function tokenMetadataErrorMode(): RosettaTokenMetadataErrorMode {
+  switch (process.env['STACKS_API_TOKEN_METADATA_ERROR_MODE']) {
+    case 'error':
+      return RosettaTokenMetadataErrorMode.error;
+    default:
+      return RosettaTokenMetadataErrorMode.warning;
+  }
+}
+
+/**
+ * LRU cache that keeps `RosettaFtMetadata` entries for FTs used in the Stacks chain and retrieved
+ * by the Rosetta endpoints.
+ */
+const ftMetadataCache = new LRUCache<string, RosettaFtMetadata>({ max: 5_000 });
```

Comment:
I don't think the caller functions do any de-duplication when looking up metadata for a set of asset IDs in a given block/tx, right? For example, if a block/tx contained say 100 transfers associated with a single asset, then this client would likely issue 100 concurrent lookups to the same RPC endpoints.

To handle that case, we probably want to change this to cache a `Promise<RosettaFtMetdata>`. That way the first call to `getFtMetdata` would immediate cache the pending lookup promise, and the additional calls would all await that first one. 

What do you think? I can open a PR against this one with that change.

Reply:
The change was pretty minor so I just pushed it in https://github.com/hirosystems/stacks-blockchain-api/pull/1747/commits/ac4282ef5112156c48053b060ef3fa227f20d5d8

Let me know what you think

Reply:
Looks great! Thanks for the change